Text baseline alignment issues

asturur opened the pull request:
This PR tries to improve the precision with which different text baselines are calculated.
With node-canvas the alphabetic and bottom baseline is always perfect, it works good.

top, middle, ideographic and hanging are a bit off.

The tweak is a 2 part change.
In the function `getBaselineAdjustment` we copy the pangoLayout and we swap the text with a string that uses all the extent of the selected font, this improve automatically the `middle` use case.

Then a correction factor is applied to the positions that do not seem correct, this correction factor is found with trial and error.

## AFTER FIX

### FIREFOX

![image](https://user-images.githubusercontent.com/1194048/80277053-b4af1b80-86ec-11ea-9c3b-2660e22ea3d1.png)
![image](https://user-images.githubusercontent.com/1194048/80277064-c7c1eb80-86ec-11ea-9692-421c3c000427.png)
![image](https://user-images.githubusercontent.com/1194048/80277201-b4635000-86ed-11ea-8392-0665ea45a6ac.png)


### CHROME

![image](https://user-images.githubusercontent.com/1194048/80277085-f0e27c00-86ec-11ea-9e58-e27247c8cfff.png)
![image](https://user-images.githubusercontent.com/1194048/80277089-f93ab700-86ec-11ea-81a0-993f3d03d5c8.png)
![image](https://user-images.githubusercontent.com/1194048/80277198-add4d880-86ed-11ea-80bc-4eed42c0d1b7.png)

## BEFORE FIX

### FIREFOX
![image](https://user-images.githubusercontent.com/1194048/80277239-fa201880-86ed-11ea-9a30-faf76b8f87fa.png)
![image](https://user-images.githubusercontent.com/1194048/80277247-04daad80-86ee-11ea-839f-e5cff77cc968.png)
![image](https://user-images.githubusercontent.com/1194048/80277258-0e641580-86ee-11ea-8e18-c90d3e7fc1da.png)

### CHROME
![image](https://user-images.githubusercontent.com/1194048/80277275-263b9980-86ee-11ea-9684-94b94111c5de.png)
![image](https://user-images.githubusercontent.com/1194048/80277280-30f62e80-86ee-11ea-9a28-be960c8b4834.png)
![image](https://user-images.githubusercontent.com/1194048/80277286-3eabb400-86ee-11ea-8028-9bc453c53e0f.png)



- [ ] Have you updated CHANGELOG.md?

asturur committed:
meh solution

asturur committed:
meh solution

asturur committed:
meh solution

asturur commented:
I also extended the tests with differenf fonts, to show this fix CAN'T work perfectly with all fonts ( middle seems an improvement anyway );

So this is the fix with the different text string use for measurement, with NO correction factor:
![image](https://user-images.githubusercontent.com/1194048/80277708-2b4e1800-86f1-11ea-8500-a76c82e52af3.png)

and this is with ALSO the correction factor
![image](https://user-images.githubusercontent.com/1194048/80277720-402aab80-86f1-11ea-8009-25a56f00d6fe.png)

asturur committed:
new test

asturur committed:
remove extra arg

asturur committed:
remove extra arg

asturur committed:
try

asturur committed:
changelog

asturur commented on test/canvas.test.js:
```diff
@@ -947,7 +947,7 @@ describe('Canvas', function () {
       assert.ok(metrics.alphabeticBaseline > 0) // ~4-5
       assert.ok(metrics.actualBoundingBoxAscent > 0)
       // On the baseline or slightly above
-      assert.ok(metrics.actualBoundingBoxDescent <= 0)
```
i tweaked the baseline offset, so i expect some text measurement to change slightly. Unsure if this is ok or not. Any thought anyone?

asturur commented on src/CanvasRenderingContext2d.cc:
```diff
@@ -2684,8 +2692,10 @@ NAN_METHOD(Context2d::MeasureText) {
+  // are those two line useful?
   cairo_matrix_t matrix;
   cairo_get_matrix(ctx, &matrix);
+
```
matrix does not seem to be used in the function. Is `cairo_get_matrix` having some side effect that is necessary for the code below to work?

chearon replied:
It looks like it is used for `alphabeticBaseline` below

asturur commented:
@zbjornson @LinusU any comment / recomendation?

LinusU commented:
This is way outside my area of expertise 😅 

I'm always a bit afraid of magic numbers, but since this clearly makes it better I think it's a good change 👍 

Hopefully @zbjornson & @chearon will have some better input 😄

chearon commented:
You should be able to get the right X-height via [pango_font_metrics_get_strikethrough_position](https://developer.gnome.org/pango/stable/pango-Fonts.html#pango-font-metrics-get-strikethrough-position) (for the "middle" baseline).

For the other ones I guess some kind of hack is reasonable. I don't quite understand it yet and I'll try to give a fuller review later. Ideally we would extract the FreeType face and look at ideographic/hanging/etc from the font tables. But Pango doesn't use FreeType on macOS/Windows so there'd have to be platform-specific code. More reasons I should remove Pango and use lower-level libraries...

asturur commented:
i m willing to write more code if it does not hinder performances.
Totally looking at pango_font_metrics_get_striketrough_position asap and remove the `different text` hack i made if it works, definitely a cleaner approach.

asturur commented:
sorry i forgot about this PR completly.

LinusU commented:
@asturur no worries, if you feel that you don't have time to improve further, I'm open to merging it as is! Thank you for spending time on this project ☺️

asturur commented:
I'll give a stab tomorrow!

asturur commented:
@charon, i tried this:

```c
  PangoRectangle logical_rect;
  PangoLayout* measureLayout = pango_layout_copy(layout);
  PangoFontMetrics *metrics;
  pango_layout_set_text(measureLayout, "gjĮ測試ÅÊ", -1);
  pango_layout_line_get_extents(pango_layout_get_line(measureLayout, 0), NULL, &logical_rect);
  metrics = PANGO_LAYOUT_GET_METRICS(measureLayout);
  int strike_thickness = pango_font_metrics_get_strikethrough_thickness(metrics);
  int strike_position = pango_font_metrics_get_strikethrough_position(metrics);
  double scale = 1.0 / PANGO_SCALE;
  double ascent = scale * pango_layout_get_baseline(measureLayout);
  double descent = scale * logical_rect.height - ascent;
  double middle = ascent - (scale * strike_position) + (scale * strike_thickness / 2);
  // 0.072 is a constant that has been chosen comparing the canvas output
  // if some code change, this constant can be changed too to keep results aligned
  double correction_factor = scale * logical_rect.height * 0.072;

```

then returned middle when reqested, but the results are far worse.

with strike throught:
![image](https://user-images.githubusercontent.com/1194048/87851952-02a77a00-c8fe-11ea-9c4f-f86fbf8ac86e.png)

with the standard ascent+descent / 2


Do you know how to better interpret that number?

Also i think i have to try on my old macbook on this new one i see that i have the text scaling bug even after recompiling.

asturur commented:
no as not said.
In the case in which the context is vertically scaled this thing breaks, Probably the striketrhoug position contains the scale factor of the context inside, and i have no idea how to revert it.

asturur committed:
Merge branch 'master' into issue-1544-textscaling

asturur commented:
The tests aren't passing anymore. While before they passed. What could it be? is something changed in tests? I reverted and push forced to the passing commit.

chearon commented:
So this PR should be updated to use `pango_font_metrics_get_strikethrough_position` for middle baseline, right?

asturur commented:
i did update it, but the results with scaling where compleltely off, so i reverted back.

```
  PangoRectangle logical_rect;
  PangoLayout* measureLayout = pango_layout_copy(layout);
  PangoFontMetrics *metrics;
  pango_layout_set_text(measureLayout, "gjĮ測試ÅÊ", -1);
  pango_layout_line_get_extents(pango_layout_get_line(measureLayout, 0), NULL, &logical_rect);
  metrics = PANGO_LAYOUT_GET_METRICS(measureLayout);
  int strike_thickness = pango_font_metrics_get_strikethrough_thickness(metrics);
  int strike_position = pango_font_metrics_get_strikethrough_position(metrics);
  double scale = 1.0 / PANGO_SCALE;
  double ascent = scale * pango_layout_get_baseline(measureLayout);
  double descent = scale * logical_rect.height - ascent;
  double middle = ascent - (scale * strike_position) + (scale * strike_thickness / 2);
  // 0.072 is a constant that has been chosen comparing the canvas output
  // if some code change, this constant can be changed too to keep results aligned
  double correction_factor = scale * logical_rect.height * 0.072;
```

The result was OK for middle when used at scale 1, but bad when used with any other scaling.
So i removed it.

BlueGhostGH commented:
Any updates on this? This is an issue that I'd really like to see fixed.

asturur commented:
i did not have time to investigate why the test started failing, i can try again rebasing the pr on a fresh branch

asturur commented:
This was a minor fix, wasn't a final solution for 1:1 match with canvas.

chearon commented:
This is still an issue but since #2602 we no longer use Pango. Let's start over with the lower-level APIs in a new PR, it should be a lot easier.

asturur commented:
oh wow i want to upgrade fabric to use version 4. on it.

chearon commented:
I would love to get your feedback! Still working on the changelog and migration guide.
